code/modules/cm_aliens/XenoStructures.dm

```diff
@@ -844,6 +845,275 @@
+
+/obj/effect/alien/resin/king_cocoon/proc/announce_halfway()
+	timer = addtimer(CALLBACK(src, PROC_REF(choose_candidate)), 4 MINUTES, TIMER_UNIQUE|TIMER_STOPPABLE|TIMER_DELETE_ME)
+
+	marine_announcement("ALERT.\n\nUNSUAL ENERGY BUILDUP DETECTED IN [get_area_name(loc)].\n\nESTIMATED TIME UNTIL COMPLETION - 5 MINUTES.", "[MAIN_AI_SYSTEM] Biological Scanner", 'sound/misc/notice1.ogg')
+	var/datum/hive_status/hive
+	for(var/cur_hive_num in GLOB.hive_datum)
+		hive = GLOB.hive_datum[cur_hive_num]
+		if(!length(hive.totalXenos))
+			continue
+		if(cur_hive_num == hive_number)
+			xeno_announcement(SPAN_XENOANNOUNCE("The King will hatch in approximately 5 minutes."), cur_hive_num, XENO_GENERAL_ANNOUNCE)
+		else
+			xeno_announcement(SPAN_XENOANNOUNCE("Another hive's King will hatch in approximately 5 minutes."), cur_hive_num, XENO_GENERAL_ANNOUNCE)
+
+#define KING_PLAYTIME_HOURS (50 HOURS)
+
+/obj/effect/alien/resin/king_cocoon/proc/is_candidate_valid(datum/hive_status/hive, mob/candidate, playtime_restricted = TRUE, skip_playtime=TRUE)
+	if(!candidate.client)
+		return FALSE
+	if(isfacehugger(candidate) || islesserdrone(candidate))
+		return FALSE
+	if(playtime_restricted)
+		if(candidate.client.get_total_xeno_playtime() < KING_PLAYTIME_HOURS)
+			return FALSE
+	else if(candidate.client.get_total_xeno_playtime() >= KING_PLAYTIME_HOURS && skip_playtime)
+		return FALSE // We do this under the assumption we tried it the other way already so don't ask twice
+	for(var/mob_name in hive.banished_ckeys)
+		if(hive.banished_ckeys[mob_name] == candidate.ckey)
+			return FALSE
+	return TRUE
+
+/obj/effect/alien/resin/king_cocoon/proc/try_roll_candidate(datum/hive_status/hive, mob/candidate, playtime_restricted = TRUE)
+	if(!is_candidate_valid(hive, candidate, playtime_restricted))
+		return FALSE
+	
+	return tgui_alert(candidate, "Would you like to become the King?", "Choice", list("Yes", "No"), 10 SECONDS) == "Yes"
+
+#undef KING_PLAYTIME_HOURS
+
+/obj/effect/alien/resin/king_cocoon/proc/make_vote(mob/candidate, list/mob/living/carbon/xenomorph/voting_candidates)
+	var/mob/living/carbon/xenomorph/choice = tgui_input_list(candidate, "Vote for a sister you wish to become the King.", "Choose a xeno", voting_candidates , 20 SECONDS)
+
+	if(votes[choice])
+		votes[choice] += 1
+	else
+		votes[choice] = 1
+
+/obj/effect/alien/resin/king_cocoon/proc/roll_candidates()
+	var/datum/hive_status/hive = GLOB.hive_datum[hive_number]
+	
+	// First, let the living xenos choose
+	var/list/mob/living/carbon/xenomorph/voting_candidates = hive.totalXenos.Copy() - hive.living_xeno_queen
+
+	for(var/mob/living/carbon/xenomorph/voting_candidate in voting_candidates)
+		if(!is_candidate_valid(hive, voting_candidate))
+			voting_candidates -= voting_candidate
+
+	for(var/mob/living/carbon/xenomorph/candidate in hive.totalXenos)
+		if(is_candidate_valid(hive, candidate, FALSE, FALSE))
+			INVOKE_ASYNC(src, PROC_REF(make_vote), candidate, voting_candidates)
+	
+	sleep(20 SECONDS)
```

Comment:
This should ideally be a timer callback e.g. roll_candidates (if you were to rename roll_candidate to start_vote since thats all it would do then); you'll just need to pass the `voting_candidates` list as an argument